deoxys/src/modes.rs

```diff
@@ -327,38 +349,49 @@ where
     }
 
     fn encrypt_decrypt_message(
-        buffer: &mut [u8],
-        tweak: &mut [u8; 16],
-        subkeys: &Array<[u8; 16], B::SubkeysSize>,
+        buffer: InOutBuf<'_, '_, u8>,
+        tweak: &mut Tweak,
+        subkeys: &Array<DeoxysKey, B::SubkeysSize>,
         tag: &Array<u8, U16>,
         nonce: &Array<u8, U15>,
     ) {
-        if !buffer.is_empty() {
-            tweak.copy_from_slice(tag);
-            tweak[0] |= 0x80;
-
-            for (index, data) in buffer.chunks_mut(16).enumerate() {
-                let index_array = (index as u64).to_be_bytes();
+        macro_rules! encrypt_decrypt_block {
+            ($index: ident, $xor: expr) => {
+                let index_array = ($index as u64).to_be_bytes();
 
                 // XOR in block numbers
                 for (t, i) in tweak[8..].iter_mut().zip(&index_array) {
                     *t ^= i
                 }
 
-                let mut block = [0u8; 16];
+                let mut block = Block::default();
                 block[1..].copy_from_slice(nonce);
 
-                B::encrypt_in_place(&mut block, tweak, subkeys);
+                B::encrypt_inout((&mut block).into(), tweak.as_ref(), subkeys);
 
-                for (t, b) in data.iter_mut().zip(block.iter()) {
-                    *t ^= b;
-                }
+                $xor(&block);
 
                 // XOR out block numbers
                 for (t, i) in tweak[8..].iter_mut().zip(&index_array) {
                     *t ^= i
                 }
+            };
+        }
+
+        if !buffer.is_empty() {
```

Comment:
To prevent drift to the right, it's worth to write this instead:
```rust
if buffer.is_empty() {
    return;
}
```

Same in the `authenticate_message` method.